api: Send user suspension email automatically + suspend all streams

victorges opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.com/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->

**What does this pull request do? Explain your changes. (required)**
This is to send the suspension email automatically when we suspend the user.
Callers only need to update their calls to include the `emailTemplate: copyright` field.

In the end also realized it was easy to add a suspension to all the user streams as well.
It won't work 100%, but it will make bans from the admin dashboard already a tad more
reliable.

**Specific updates (required)**
 - Send suspension email when user is suspended
 - Suspend and terminate all streams of the user when the user is suspended

## -

- **How did you test each of these updates (required)**
Trust the code.

**Does this pull request close any open issues?**
Implements #987 

**Checklist:**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

victorges committed:
api: Create schema for suspend user API

victorges committed:
api: Send account suspension email on user suspended

victorges committed:
bonus: Suspend all user streams on suspension

gioelecerati commented on packages/api/src/controllers/user.ts:
```diff
@@ -60,6 +62,39 @@ async function findUserByEmail(email: string, useReplica = true) {
+  path: string
+) => `${proto || "http"}://${frontendDomain}${path}`;
+
+const unsubscribeUrl = (req: Request) => frontendUrl(req, "/contact");
```
This is also in the other sendGrid notification pr we merged, we should at some point put a correct unsubscribe url for `{{{unsubscribe}}}` in the sendGrid template? Or at least not use an "unsubscribe" link into the email template but a "contact us" one, as it's a notification for the service and not a newsletter

victorges replied:
Yeah... we need a real unsubscribe URL at some point I guess (tho still weird cause it's not a newsletter. we need this communication access and if you don't like the webhooks just disable them)

victorges committed:
api/user: Log with logger instead of console

Didn't even know we had that

victorges committed:
api: Decouple appeal button from unsubscribe

gioelecerati commented on packages/api/src/controllers/user.ts:
```diff
@@ -60,6 +62,39 @@ async function findUserByEmail(email: string, useReplica = true) {
+  suspended: boolean
+): Promise<void> {
+  const [streams] = await db.stream.find({ userId });
+  for (const stream of streams) {
```
Nice! Once for all of them

victorges replied:
Hopefully. The only problem I guess is that it's not really reliable, which is why we need that MistUtilNukeProc or something like that, running directly on the container :/

But maybe the problem is just something on our side, idk.

gioelecerati commented on packages/api/src/controllers/user.ts:
```diff
@@ -326,23 +355,55 @@ app.post("/", validatePost("user"), async (req, res) => {
+  validatePost("suspend-user-payload"),
   authMiddleware({ anyAdmin: true }),
   async (req, res) => {
+    const { suspended, emailTemplate } = req.body as SuspendUserPayload;
```
Just a question, why is the emailTemplate sent as a body parameter instead of being defined somewhere in livepeer-com? I think the only one is `copyright` anyway, right?

victorges replied:
Yeah but maybe we don't want to send the email when we suspend, so the template is optional. And later/soon we might have other reasons to suspend the user so I thought it made sense to be more than just an opt-in boolean here.

gioelecerati commented on packages/api/src/controllers/user.ts:
```diff
@@ -326,23 +355,55 @@ app.post("/", validatePost("user"), async (req, res) => {
+      } = req.config;
+      try {
+        await sendgridEmail({
+          email,
```
Should we CC also infraservice as we are manually doing right now?

victorges replied:
Ah, right, forgot about that

gioelecerati commented on packages/api/src/controllers/user.ts:
```diff
@@ -60,6 +62,39 @@ async function findUserByEmail(email: string, useReplica = true) {
   return users[0];
 }
 
+const frontendUrl = (
```
Nice! Maybe `buttonUrl = (` as it's always used for the `buttonUrl` field?

victorges replied:
I actually had `emailUrl` since they're always included in emails. I don't think that makes a lot of sense tho since you would then have to look at where this is used to make sense out of it, instead of it making sense on its own. Then I thought that the most appropriate name would be this, since what it really does is create a URL pointing to a frontend page.

gioelecerati reviewed (commented):
Adding the pending comments before losing them

gioelecerati commented on packages/api/src/controllers/user.ts:
```diff
@@ -326,23 +355,55 @@ app.post("/", validatePost("user"), async (req, res) => {
+      const {
+        frontendDomain,
+        supportAddr,
+        sendgridTemplateId,
```
Same as above, we have the `sendgridTemplateId` defined anyway, so only copyright with this Id is possible

victorges replied:
Hmm, we have a template for the outer portion of the email, but we still need to provide the content for it including the subject, text, etc, right? So that's what I meant with the copyright "email template". We can rethink any of this later when we either add more sendgrid templates or more "text templates", I guess.

I don't think automatically sending an email about copyright infringiment makes sense tho, so that's why I wanted to specify what email is supposed to be sent. There are various reasons why we might want to suspend an user, and forgetting that it automatically sends a copyright email and sending it when it was not the case seems worse to me than forgetting about including the param here and not sending the email sometimes (which won't happen with the automatic tooling anyway).

victorges replied:
OK I changed the code a bit to make my intent clearer, and also made sure that we always email users, using just a simple message when the copyright is not specified.

LMK what you think!
https://github.com/livepeer/livepeer-com/pull/989/commits/1d1d529ea4c38e2cd4a8d11b52138eecc9012256

gioelecerati replied:
Oh ok I see! Much clearer and makes sense!

victorges replied:
Default email template powered by GitHub Autopilot®! 😂

gioelecerati reviewed (approved):
LGTM

victorges committed:
api: Add bcc to suspension email

victorges committed:
api: Create suspensionEmailText helper

Also always email the user with some simple template,
but use the copyright one if specified.

victorges committed:
Fix tests
